KAFKA-3272: Add debugging options to kafka-run-class.sh so we can easily run remote debugging

christian-posta opened the pull request:
…able remote debugging to Kafka tools scripts

christian-posta committed:
Fix for https://issues.apache.org/jira/browse/KAFKA-3272 to easily enable remote debugging to Kafka tools scripts

gwenshap commented:
Yes! Love this. Maybe we need a flag for suspend=y? I find that I need it more often than not.

christian-posta commented:
yah good point @gwenshap! tbh, i always use that flag. let me add it as an option.

granthenke commented on bin/kafka-run-class.sh:
```diff
@@ -115,6 +115,23 @@ if [ -z "$KAFKA_OPTS" ]; then
 
+# Set Debug options if enabled
+DEFAULT_JAVA_DEBUG_PORT="5005"
+if [ "x$JAVA_DEBUG_PORT" = "x" ]; then
```
Can we change this to match the other "presence" checks in the script:

```
if [ -z "$JAVA_DEBUG_PORT" ]
```

granthenke replied:
Should this move inside the section below  the `if KAFKA_DEBUG` check?

christian-posta replied:
yah let me refactor that, thanks for the review!

granthenke commented on bin/kafka-run-class.sh:
```diff
@@ -115,6 +115,23 @@ if [ -z "$KAFKA_OPTS" ]; then
+    JAVA_DEBUG_PORT="$DEFAULT_JAVA_DEBUG_PORT"
+fi
+
+DEFAULT_JAVA_DEBUG_OPTS="-agentlib:jdwp=transport=dt_socket,server=y,suspend=${DEBUG_SUSPEND_FLAG:-n},address=$JAVA_DEBUG_PORT"
```
Should this move inside the section below the `if KAFKA_DEBUG` check?

christian-posta replied:
+1 will change

granthenke commented on bin/kafka-run-class.sh:
```diff
@@ -115,6 +115,23 @@ if [ -z "$KAFKA_OPTS" ]; then
+fi
+
+DEFAULT_JAVA_DEBUG_OPTS="-agentlib:jdwp=transport=dt_socket,server=y,suspend=${DEBUG_SUSPEND_FLAG:-n},address=$JAVA_DEBUG_PORT"
+if [ "x$KAFKA_DEBUG" != "x" ]; then
```
Same as above, use `-z`

granthenke commented on bin/kafka-run-class.sh:
```diff
@@ -115,6 +115,23 @@ if [ -z "$KAFKA_OPTS" ]; then
+DEFAULT_JAVA_DEBUG_OPTS="-agentlib:jdwp=transport=dt_socket,server=y,suspend=${DEBUG_SUSPEND_FLAG:-n},address=$JAVA_DEBUG_PORT"
+if [ "x$KAFKA_DEBUG" != "x" ]; then
+    # Use the defaults if JAVA_DEBUG_OPTS was not set
+    if [ "x$JAVA_DEBUG_OPTS" = "x" ]; then
```
Same as above, use `-z`

granthenke commented:
Thanks for the patch, just a few small comments. 

Could you change the title of the PR to: "KAFKA-3272: Add debugging options to kafka-run-class.sh so we can easily run remote debugging" so that it links to the jira?

Would you be interested in making the Windows .bat script match too?

christian-posta commented:
@granthenke force-updated the commit with the changes you suggested. thanks!

let me poke around at the windows script.

granthenke commented on bin/kafka-run-class.sh:
```diff
@@ -115,6 +115,23 @@ if [ -z "$KAFKA_OPTS" ]; then
+    fi
+
+    KAFKA_OPTS="$JAVA_DEBUG_OPTS $KAFKA_OPTS"
+    warn "Enabling Java debug options: $JAVA_DEBUG_OPTS"
```
Not sure you need the `warn` line

gwenshap commented:
@granthenke @christian-posta Do you think we should merge now and follow up on windows later? Or add windows to this PR and avoid diverging the scripts?

granthenke commented on bin/kafka-run-class.sh:
```diff
@@ -115,6 +115,28 @@ if [ -z "$KAFKA_OPTS" ]; then
+
+
+    KAFKA_OPTS="$JAVA_DEBUG_OPTS $KAFKA_OPTS"
+    warn "Enabling Java debug options: $JAVA_DEBUG_OPTS"
```
I am not familiar with `warn`. Do we need this?

granthenke commented:
@gwenshap I am okay, if we merge and open a jira for windows. I just had one comment on the usage of `warn`. Otherwise LGTM.

christian-posta committed:
added windows scripts for https://issues.apache.org/jira/browse/KAFKA-3272 to easily enable remote debugging to Kafka tools scripts

christian-posta commented:
@granthenke @gwenshap i've changed the warn to just "echo" and added the windows equivalent. 

windows batch file expansion rules.. learn something new everyday!

gwenshap commented:
...if there was only a place to document those cool new features :)

Actually maybe we don't have a good place for that yet. I'll commit and open a JIRA for command line docs.

gwenshap commented:
LGTM.

christian-posta commented:
@gwenshap shall I open another JIRA for the patch for the windows batch files? Already have it in this branch...
